### ISSUE TITLE
`ML_estimate_component_based_normalisation` and `apply_norm_factors3D` cannot be used without saving to disk

### ISSUE THREAD

**Issue by robbietuk** [UCL/STIR#1498]
These are executable files for computing normalization and a brief description of how they are implmented:
- [`find_ML_normfactors3D`](https://github.com/UCL/STIR/blob/master/src/utilities/find_ML_normfactors3D.cxx): calls [`ML_estimate_component_based_normalisation`](https://github.com/UCL/STIR/blob/master/src/recon_buildblock/ML_estimate_component_based_normalisation.cxx) and the single function there. Additionally, the documentation in [`ML_estimate_component_based_normalisation.h`](https://github.com/UCL/STIR/blob/master/src/include/stir/recon_buildblock/ML_estimate_component_based_normalisation.h) is lacking.
- [`find_ML_normfactors`](https://github.com/UCL/STIR/blob/master/src/utilities/find_ML_normfactors.cxx): implements much of the functionality in the main function.
- [`apply_normfactors3D`](https://github.com/UCL/STIR/blob/master/src/utilities/apply_normfactors3D.cxx): implements things in the main function, at least is very short.
- [`apply_normfactors`](https://github.com/UCL/STIR/blob/master/src/utilities/apply_normfactors.cxx): does the same as `apply_normfactors3D` _(code duplication?)_

Most of the following pertains to `find_ML_normfactors3D` and `find_ML_normfactors`. The above functionality saves the computed data to be saved to disk, rather than returning/utilizing a shared_ptr.

__________

The core functionality of these executables should be split into various components.
1. The main function should handle only the arguments handling and saving data.
1. Mostly, functions/classes that take the core arguments and compute/process the data.
1. Can handle saving of the data, if specified by the arguments.
1. Function/ object returns a pointer to the computed data, that can be saved by the main exe.

_____

To implement this:
- The exe's need to maintain backwards compatibility.
- Replace the `ML_estimate_component_based_normalisation` function with a class/function.. 
- Allow member variables/argument to determine if data is saved to disk.
- Store/return computed data as members that can be saved/handled by the exe main functions.

**Comment by robbietuk** [comment]
Upon further review of [`apply_normfactors3D`](https://github.com/UCL/STIR/blob/master/src/utilities/apply_normfactors3D.cxx), the utility does in fact use the [`BinNormalisationPETFromComponents`](https://github.com/UCL/STIR/blob/master/src/include/stir/recon_buildblock/BinNormalisationPETFromComponents.h) class. However, there is currently no setter for the `crystal_efficiencies`, `geometric_factors` and `block_factors`as the utility uses a streaming method to get the data
https://github.com/UCL/STIR/blob/c51067d72efe914d8a344fac913ffc9db0fdba39/src/utilities/apply_normfactors3D.cxx#L91-L97

**Comment by KrisThielemans** [comment]
Yes, the intention was to let all this work via `BinNormalisationPETFromComponents`. The current file IO is rather ugly and inconvenient, but we'll need to have it for backwards compatibility for a while. In principle, `ML_estimate...` could directly return a `BinNormalisationPETFromComponents` object. That needs some careful thought however.

**Comment by robbietuk** [comment]
> Yes, the intention was to let all this work via `BinNormalisationPETFromComponents`.

Do you mean for `BinNormalisationPETFromComponents` to have functionality to perform `ML_estimate...` and compute the `efficiencies` etc?

> In principle, `ML_estimate...` could directly return a `BinNormalisationPETFromComponents` object. That needs some careful thought however.

It could be as simple as a method on the `ML_estimate...` class having a method that constructs a `BinNormalisationPETFromComponents`. I think I am confused as to why architecture is such that the `find_ML_normfactors3D` and `apply_normfactors3D` utilities are split.

**Comment by KrisThielemans** [comment]
> > Yes, the intention was to let all this work via `BinNormalisationPETFromComponents`.
> 
> Do you mean for `BinNormalisationPETFromComponents` to have functionality to perform `ML_estimate...` and compute the `efficiencies` etc?
>

no, I don't think that's a good idea.
 
> > In principle, `ML_estimate...` could directly return a `BinNormalisationPETFromComponents` object. That needs some careful thought however.
> 
> It could be as simple as a method on the `ML_estimate...` class having a method that constructs a `BinNormalisationPETFromComponents`. 

yes

> I think I am confused as to why architecture is such that the `find_ML_normfactors3D` and `apply_normfactors3D` utilities are split.

well, `Reconstruction` is a different hierarchy than `DiscretisedDensity` :-). Why would we put "estimation" and "usage" in one class?

**Comment by robbietuk** [comment]
> > > Yes, the intention was to let all this work via `BinNormalisationPETFromComponents`.
> > 
> > 
> > Do you mean for `BinNormalisationPETFromComponents` to have functionality to perform `ML_estimate...` and compute the `efficiencies` etc?
> 
> no, I don't think that's a good idea.
> 
> > > In principle, `ML_estimate...` could directly return a `BinNormalisationPETFromComponents` object. That needs some careful thought however.
> > 
> > 
> > It could be as simple as a method on the `ML_estimate...` class having a method that constructs a `BinNormalisationPETFromComponents`.
> 
> yes
> 
> > I think I am confused as to why architecture is such that the `find_ML_normfactors3D` and `apply_normfactors3D` utilities are split.
> 
> well, `Reconstruction` is a different hierarchy than `DiscretisedDensity` :-). Why would we put "estimation" and "usage" in one class?
